### PR TITLE
added support for go-sdk-core authenticator

### LINKED ISSUE #93: Add support for the github.com/IBM/go-sdk-core authenticator type
Most of the IBM Cloud clients are using the github.com/IBM/go-sdk-core authenticators for authenticating to the IBM Cloud, so even this client should start using the same package for going forward to comply with the other clients.

e.g: how to use for apikey: https://github.com/IBM/go-sdk-core/blob/main/Authentication.md#identity-and-access-management-authentication-iam

### PR DESCRIPTION
**Pull request by Prajyot-Parab**
Fixes #93 
Signed-off-by: Prajyot-Parab <prajyot.parab2@ibm.com>

### CODE REVIEW

**Comment by Prajyot-Parab** [comment]
@yussufsh PTAL

**Review by mkumatag — APPROVED** [review]
lgtm

File: ibmpisession/ibmpowersession.go [hunk 1]
```diff
@@ -76,16 +99,60 @@ region : Obtained from the terraform template. Every template /resource will be
...
+This method will create a IBMPISession based on the options passed by client.
+It will pass the newly created IBMPISession to newSubProcess method which will evaluate and set rest of the required options from session.
+*/
+func NewSession(options *PIOptions) (*IBMPISession, error) {
+	t := timeout
+	if options.Timeout.Seconds() != 0 {
+		t = options.Timeout
+	}
+
+	newSession := &IBMPISession{
```
Comment (yussufsh) [inline ibmpisession/ibmpowersession.go:131]: You are not using `IAMToken` in the new struct. Deprecating the `New` function should not mean that we won't allow auth via direct token.
This is currently in use by the Terraform provider code which is the main use case for this project.

Comment (mkumatag) [reply ibmpisession/ibmpowersession.go:131]: `NewSession` this function itself can generate that token you are talking about, user no need to generate the token in advance. Even tf can start using this function eventually.

File: ibmpisession/ibmpowersession.go [hunk 2]
```diff
@@ -99,13 +166,28 @@ func New(iamtoken, region string, debug bool, timeout time.Duration, useraccount
 // NewAuth ...
 func NewAuth(sess *IBMPISession, PowerInstanceID string) runtime.ClientAuthInfoWriter {
 	var crndata = crnBuilder(PowerInstanceID, sess.UserAccount, sess.Region, sess.Zone)
+	var authdata = sess.IAMToken
 	return runtime.ClientAuthInfoWriterFunc(func(r runtime.ClientRequest, _ strfmt.Registry) error {
-		if err := r.SetHeaderParam("Authorization", sess.IAMToken); err != nil {
+
```
Comment (yussufsh) [inline ibmpisession/ibmpowersession.go:171]: Please move below code to a helper func and we need to also use it in `BearerTokenAndCRN` func as well.

**Review by yussufsh — CHANGES_REQUESTED** [review]

File: ibmpisession/ibmpowersession.go [hunk 3]
```diff
@@ -99,13 +166,28 @@ func New(iamtoken, region string, debug bool, timeout time.Duration, useraccount
...
+		/*
+			Check if authenticator option is set
+			If yes, authenticate using it and fetch authorization header data required for the request.
+		*/
+		if sess.Authenticator != nil {
+			req := &http.Request{
+				Header: make(http.Header),
+			}
+			if err := sess.Authenticator.Authenticate(req); err != nil {
```
[the remark this answers is not in the document]
Comment (mkumatag) [reply ibmpisession/ibmpowersession.go:180]: Most of the authenticators do call this function while creating the authenticator object itself and also while generating the token, having that said, no harm in calling that explicitly here.

**Review by yussufsh — APPROVED** [review]

**Pull request merged**